### ISSUE TITLE
sbt 1.1.x watch (~) thread leak causing high cpu use

### ISSUE THREAD

**Issue by francisdb** [sbt/sbt#3860]
sbt version: 1.1.0/1.1.1
MacOS 10.13.2
java version "1.8.0_152"
Java(TM) SE Runtime Environment (build 1.8.0_152-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.152-b16, mixed mode)

We have a big project (40 subprojects)

## steps

```
sbt
~compile
[enter]
~compile
[enter]
~compile
[enter]
...
```
(on the root project or even on a subproject)

## problem

cpu use goes up a percentage on every `~compile` and does not go down after `[enter]`, in the end all my cpu (8 cores) is in use by the sbt java process. Only after exiting sbt cpu goes back to normal.

## expectation

cpu falls back to a low percentage after ending the watch

## notes

My impression is that the watchers never shut down. Is there anything I can do to debug this?

might be related to #3820 even if I have not seen that issue today

This is a thread view in jvisualvm (looks like something should be using a named thread factory)

<img width="1080" alt="image" src="https://user-images.githubusercontent.com/161305/34676041-ba1b148e-f48b-11e7-8d93-5ef77cdb445a.png">

This is part of the thead dump (focusing on these unnamed threads)

```
"Thread-1543" #4419 prio=5 os_prio=31 tid=0x00007fb763eae000 nid=0x8cdf runnable [0x000070000c8cb000]
   java.lang.Thread.State: RUNNABLE
        at java.io.UnixFileSystem.getBooleanAttributes0(Native Method)
        at java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:242)
        at java.io.File.isDirectory(File.java:849)
        at sbt.io.DirectoryFilter$.accept(NameFilter.scala:84)
        at java.io.File.listFiles(File.java:1291)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1(Path.scala:437)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1$adapted(Path.scala:435)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1357/1174641185.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.DescendantOrSelfPathFinder.addTo(Path.scala:435)
        at sbt.io.PathFinder.get(Path.scala:380)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2(PollingWatchService.scala:96)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2$adapted(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread$$Lambda$4246/1728697756.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.PollingWatchService$PollingThread.getFileTimes(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread.populateEvents(PollingWatchService.scala:110)
        at sbt.io.PollingWatchService$PollingThread.run(PollingWatchService.scala:86)

   Locked ownable synchronizers:
        - None

"Thread-1433" #3994 prio=5 os_prio=31 tid=0x00007fb7624a8000 nid=0x9f1f runnable [0x000070000c7c9000]
   java.lang.Thread.State: RUNNABLE
        at com.sun.jna.Native.malloc(Native Method)
        at com.sun.jna.Memory.malloc(Memory.java:721)
        at com.sun.jna.Memory.<init>(Memory.java:113)
        at com.sun.jna.NativeString$StringMemory.<init>(NativeString.java:42)
        at com.sun.jna.NativeString.<init>(NativeString.java:92)
        at com.sun.jna.NativeString.<init>(NativeString.java:65)
        at com.sun.jna.Function.convertArgument(Function.java:557)
        at com.sun.jna.Function.invoke(Function.java:338)
        at com.sun.jna.Library$Handler.invoke(Library.java:244)
        at com.sun.proxy.$Proxy11.getattrlist(Unknown Source)
        at sbt.internal.io.MacMilli$.$anonfun$getModifiedTimeNative$3(Milli.scala:246)
        at sbt.internal.io.MacMilli$$$Lambda$881/1199155942.apply$mcI$sp(Unknown Source)
        at sbt.internal.io.MilliPosixBase.checkedIO(Milli.scala:93)
        at sbt.internal.io.MacMilli$.getModifiedTimeNative(Milli.scala:246)
        at sbt.internal.io.MacMilli$.getModifiedTimeNative(Milli.scala:239)
        at sbt.internal.io.MilliNative.getModifiedTime(Milli.scala:66)
        at sbt.internal.io.Milli$.getModifiedTime(Milli.scala:352)
        at sbt.io.IO$.getModifiedTime(IO.scala:1143)
        at sbt.io.IO$.getModifiedTimeOrZero(IO.scala:1212)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$3(PollingWatchService.scala:96)
        at sbt.io.PollingWatchService$PollingThread$$Lambda$4247/1664507301.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2(PollingWatchService.scala:96)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2$adapted(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread$$Lambda$4246/1728697756.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.PollingWatchService$PollingThread.getFileTimes(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread.populateEvents(PollingWatchService.scala:110)
        at sbt.io.PollingWatchService$PollingThread.run(PollingWatchService.scala:86)

   Locked ownable synchronizers:
        - None

"Thread-1323" #3592 prio=5 os_prio=31 tid=0x00007fb752002800 nid=0x65ab runnable [0x000070000c6c6000]
   java.lang.Thread.State: RUNNABLE
        at java.io.UnixFileSystem.list(Native Method)
        at java.io.File.list(File.java:1122)
        at java.io.File.listFiles(File.java:1286)
        at sbt.io.FilterFiles.handleFile(Path.scala:428)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:442)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1(Path.scala:437)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1$adapted(Path.scala:435)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1357/1174641185.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.DescendantOrSelfPathFinder.addTo(Path.scala:435)
        at sbt.io.PathFinder.get(Path.scala:380)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2(PollingWatchService.scala:96)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2$adapted(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread$$Lambda$4246/1728697756.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.PollingWatchService$PollingThread.getFileTimes(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread.populateEvents(PollingWatchService.scala:110)
        at sbt.io.PollingWatchService$PollingThread.run(PollingWatchService.scala:86)

   Locked ownable synchronizers:
        - None

"Thread-1213" #3197 prio=5 os_prio=31 tid=0x00007fb76c12c000 nid=0x6b6f runnable [0x000070000c5c3000]
   java.lang.Thread.State: RUNNABLE
        at java.io.UnixFileSystem.list(Native Method)
        at java.io.File.list(File.java:1122)
        at java.io.File.listFiles(File.java:1286)
        at sbt.io.FilterFiles.handleFile(Path.scala:428)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:442)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1(Path.scala:437)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1$adapted(Path.scala:435)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1357/1174641185.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.DescendantOrSelfPathFinder.addTo(Path.scala:435)
        at sbt.io.PathFinder.get(Path.scala:380)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2(PollingWatchService.scala:96)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2$adapted(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread$$Lambda$4246/1728697756.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.PollingWatchService$PollingThread.getFileTimes(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread.populateEvents(PollingWatchService.scala:110)
        at sbt.io.PollingWatchService$PollingThread.run(PollingWatchService.scala:86)

   Locked ownable synchronizers:
        - None

"Thread-422" #1517 prio=5 os_prio=31 tid=0x00007fb764349800 nid=0x9d23 runnable [0x000070000c4c0000]
   java.lang.Thread.State: RUNNABLE
        at java.io.UnixFileSystem.list(Native Method)
        at java.io.File.list(File.java:1122)
        at java.io.File.listFiles(File.java:1286)
        at sbt.io.FilterFiles.handleFile(Path.scala:428)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:442)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1(Path.scala:444)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$handleFileDescendant$1$adapted(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1358/1126074033.apply(Unknown Source)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at sbt.io.DescendantOrSelfPathFinder.handleFileDescendant(Path.scala:443)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1(Path.scala:437)
        at sbt.io.DescendantOrSelfPathFinder.$anonfun$addTo$1$adapted(Path.scala:435)
        at sbt.io.DescendantOrSelfPathFinder$$Lambda$1357/1174641185.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.DescendantOrSelfPathFinder.addTo(Path.scala:435)
        at sbt.io.PathFinder.get(Path.scala:380)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2(PollingWatchService.scala:96)
        at sbt.io.PollingWatchService$PollingThread.$anonfun$getFileTimes$2$adapted(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread$$Lambda$4246/1728697756.apply(Unknown Source)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at sbt.io.PollingWatchService$PollingThread.getFileTimes(PollingWatchService.scala:93)
        at sbt.io.PollingWatchService$PollingThread.populateEvents(PollingWatchService.scala:110)
        at sbt.io.PollingWatchService$PollingThread.run(PollingWatchService.scala:86)

   Locked ownable synchronizers:
        - None

```

**Comment by francisdb** [comment]
This shows the thread buildup

![image](https://user-images.githubusercontent.com/161305/34676406-013d06b4-f48d-11e7-804c-b9c519918f2b.png)

<img width="1084" alt="image" src="https://user-images.githubusercontent.com/161305/34676433-12cf8be0-f48d-11e7-8aaa-7c1ade12d2d5.png">

**Comment by olafurpg** [comment]
Have you tried using another file watcher? I had the same problems in sbt 1.0 until I enabled https://github.com/swoval/MacOSXWatchService globally, been smooth sailing since. I wish sbt 1.0 came bundled with a better macos file watcher by default, it was pretty much unusable without MacOSXWatchService for me.

**Comment by eatkins** [comment]
I think this is related to my outstanding PR: https://github.com/sbt/sbt/pull/3790.

**Comment by francisdb** [comment]
@olafurpg have been using that one before, just gave it another try but still issues for me.

**Comment by francisdb** [comment]
@eatkins any chance you could give that thread a proper name?

**Comment by eatkins** [comment]
Sure, all plugin created threads in version 1.2.0-RC3 of MacOSXWatchService should have names.

**Comment by francisdb** [comment]
sbt `1.1.0` is again usable for me with MacOSXWatchService `1.2.0-RC3`, looking forward to this being fixed in sbt `1.1.1`

(also confirming visualvm shows a single named thread)

**Comment by dwijnand** [comment]
Fixed with https://github.com/sbt/sbt/pull/3790?

**Comment by francisdb** [comment]
How do I test a snapshot of sbt?

**Comment by dwijnand** [comment]
If you use sbt-extras just `publishLocal` and then change `sbt.version` to whatever the version is.

**Comment by francisdb** [comment]
Yes the thread leak seems to be fixed.
But
 * The thread name is still `Thread-xxx`
 * When idle with `~compile` I see a lot more CPU activity (~10%) than when using MacOSXWatchService (~1%)

**Comment by francisdb** [comment]
Looks to me like this has NOT been fixed in sbt `1.1.1`. Has the previous fix been reverted?

One extra watch thread is still created per `~test:compile` start (and stop), taking up more and more cpu.

<img width="545" alt="image" src="https://user-images.githubusercontent.com/161305/36090427-a2221c18-0fe0-11e8-996b-6f81f4d33656.png">

**Comment by dwijnand** [comment]
No the fix #3790 was merged into 1.x after 1.1.0 was released so it's scheduled for sbt 1.2.0.

But if anyone fancies backporting it to the 1.1.x branch it could be part of the 1.1.2 release.

**Comment by eed3si9n** [comment]
Fixed in https://github.com/sbt/sbt/pull/4071 / sbt/io#138, and will be released in 1.1.3.
